src/main/java/com/apeer/sdk/impl/ApeerDevKit.java

```diff
@@ -0,0 +1,144 @@
+        } else if (type == Boolean.class || type == boolean.class) {
+            return (T) (Boolean) inputJson.getBoolean(key);
+        } else if (type.isArray()) {
+            var jsonItems = inputJson.getJSONArray(key);
+            T items = (T) Array.newInstance(type.getComponentType(), jsonItems.length());
+            for (int i = 0; i < jsonItems.length(); i++) {
+                Array.set(items, i, jsonItems.get(i));
+            }
+            return items;
+        } else {
+            throw new ApeerInputException("Input type \"" + type.toString() + "\" is not allowed. Use String, Integer, Double or Boolean");
+        }
+    }
+
+    /**
+     * Sets the output that will be written to output_params_file
+     *
+     * @param key   The output key as defined in the module_specification.json of your module
+     * @param value The value that will be written to the associated key. Can be any type that is supported by JSON but
+     *              must correspond to the type as specified in the module_specification.json of your module
+     * @throws ApeerOutputException When the value could not be parsed to JSON
+     */
+    public void setOutput(String key, Object value) throws ApeerOutputException {
+        try {
+            outputJson.put(key, value);
+        } catch (JSONException ex) {
+            throw new ApeerOutputException("Could not set output \"" + key + "\"", ex);
+        }
+    }
+
+    /**
+     * Sets a file output that will be written to output_params_file. Also copies the file to the output folder of your
+     * module as required by the APEER environment
+     *
+     * @param key            The output key as defined in the module_specification.json of your module
+     * @param outputFilePath The relative path to your file as you saved it
+     * @throws ApeerOutputException When the value could not be parsed to JSON
+     */
+    public void setFileOutput(String key, String outputFilePath) throws ApeerOutputException {
+        if (!outputFilePath.startsWith("/output/")) {
+            var targetPath = "/output/" + outputFilePath;
+            fileOutputWriter.moveFile(Path.of(outputFilePath), Path.of(targetPath));
+            outputFilePath = targetPath;
+        }
+
+        try {
+            outputJson.put(key, outputFilePath);
+        } catch (JSONException ex) {
+            throw new ApeerOutputException("Could not set output \"" + key + "\"", ex);
+        }
+    }
+
+    /**
+     * Writes all output values as defined via {@code setOutput} and {@code setFileOutput} to the output params file
+     *
+     * @throws ApeerOutputException When the output params file could not be written
+     */
+    public void finalizeModule() throws ApeerOutputException {
+        var json = outputJson.toString();
+        fileOutputWriter.writeTextToFile(outputParamsFile, json);
+    }
+
+    private void log(String message) {
+        System.out.println("[ADK] " + message);
```

Comment:
I dont what the convention is, but I think adk should not log information. I think looging should be a part of user's code. So that a user has full control over his output logs.

Reply:
It really help when looking at the module logs. Matlab ADK also has this. So will leave this.!